tor/controller.go

```diff
@@ -0,0 +1,471 @@
+// Tor server and the version of the Tor server.
+func (c *Controller) ProtocolInfo() ([]string, string, string, error) {
+	// We'll start off by sending the "PROTOCOLINFO" command to the Tor
+	// server. We should receive a reply of the following format:
+	//
+	//	METHODS=COOKIE,SAFECOOKIE
+	//	COOKIEFILE="/home/user/.tor/control_auth_cookie"
+	//	VERSION Tor="0.3.2.10"
+	//
+	// We're interested in retrieving all of these fields, so we'll parse
+	// our reply to do so.
+	cmd := fmt.Sprintf("PROTOCOLINFO %d", ProtocolInfoVersion)
+	_, reply, err := c.sendCommand(cmd)
+	if err != nil {
+		return nil, "", "", err
+	}
+
+	info := parseTorReply(reply)
+	methods, exists := info["METHODS"]
+	if !exists {
+		return nil, "", "", errors.New("auth methods not found in " +
+			"reply")
+	}
+
+	cookieFile, exists := info["COOKIEFILE"]
+	if !exists {
+		return nil, "", "", errors.New("cookie file path not found " +
+			"in reply")
+	}
+
+	version, exists := info["Tor"]
+	if !exists {
+		return nil, "", "", errors.New("Tor version not found in reply")
+	}
+
+	// Finally, we'll clean up the results before returning them.
+	authMethods := strings.Split(methods, ",")
+	cookieFilePath := strings.Trim(cookieFile, "\"")
+	torVersion := strings.Trim(version, "\"")
+
+	return authMethods, cookieFilePath, torVersion, nil
+}
+
+// AddOnionV2 creates a new v2 onion service and returns its onion address(es).
+// Once created, the new onion service will remain active until the connection
+// between the controller and the Tor server is closed. The path to a private
+// key can be provided in order to restore a previously created onion service.
+// If a file at this path does not exist, a new onion service will be created
+// and its private key will be saved to a file at this path. A mapping of
+// virtual ports to target ports should also be provided. Each virtual port will
+// be the ports where the onion service can be reached at, while the mapped
+// target ports will be the ports where the onion service is running locally.
+// For example, if you wish to run an onion service that should be accessed on
+// port 9735, but the service is listening on 9000, 9001, and 9002 internally,
+// then a mapping of {9735: {9001, 9002, 9003}} can be provided.
+func (c *Controller) AddOnionV2(privateKeyFile string,
+	virtToTargPorts map[int][]int) ([]*OnionAddr, error) {
+
+	// We'll start off by reading the private key from its path.
+	privateKey, err := ioutil.ReadFile(privateKeyFile)
+	if err != nil && !strings.Contains(err.Error(), "no such file") {
+		return nil, fmt.Errorf("unable to read private key from file: "+
+			"%v", err)
+	}
+
+	// If the file does not exist, then we should request the server to
+	// create a new onion service and return its private key. Otherwise,
+	// we'll request the server to recreate the onion server from our
+	// private key.
+	var keyParam string
+	if privateKey != nil {
+		keyParam = string(privateKey)
```

Comment:
Looking at the spec, this needs to be a base64 DER encoded version of the RSA private key: https://github.com/torproject/torspec/blob/master/control-spec.txt#L1625